src/search/dag_classic/search.cc

```
@@ -2225,18 +2237,42 @@ void SearchWorker::DoBackupUpdateSingleNode(
   auto path = node_to_process.path;
 
   if (node_to_process.nn_queried) {
+#ifdef FIX_TT
+    auto entry = search_->tt_->LookupAndPin(node_to_process.hash);
+    bool is_tt_miss = !entry;
+#else
     auto [tt_iter, is_tt_miss] = search_->tt_->try_emplace(
         node_to_process.hash, node_to_process.tt_low_node);
+#endif
     if (is_tt_miss) {
+#ifdef FIX_TT
+      search_->tt_->Insert(node_to_process.hash,
+                           std::make_unique<std::weak_ptr<LowNode>>(
+                               node_to_process.tt_low_node));
```

I think this needs a comment that insertions aren't thread safe if a hash is the same. Current implementation silently ignores insertion if hash is already present in the table.

I was thinking about last inserted eviction policy. It isn't perfect match how search access nodes. We can end up evicting nodes which are old insertion which are still close to the PV line. A better eviction policy would be least recently accessed but that would be expensive to implement without major refactoring.
We could reduce risk of eviction in insertion if garbage collector evicts expired weak pointers. I don't think initial version needs this. It can be an useful change when playing matches against strong engines where both agree on a long line.

The table implementation requires unique_ptr storage. Could we modify storage to support weak_ptr too? I think this is also a change that shouldn't be part of initial version. It could be a future improvement. It could be also redundant work if we choose to give ownership to the hash table to allow cycles in the tree.

---

I don't see the thread safety issue, can you elaborate? The insertion failing shouldn't cause any correctness issues.